src/module.c

```diff
@@ -118,9 +153,83 @@ static int _RenameGraphHandler(RedisModuleCtx *ctx, int type, const char *event,
 	return REDISMODULE_OK;
 }
 
-static void RegisterForkHooks(RedisModuleCtx *ctx) {
+// Module level function for RDB start save event - Create the meta keys for each graph.
+static void _CreateKeySpaceMetaKeys(RedisModuleCtx *ctx) {
+	uint graphs_in_keyspace_count = array_len(graphs_in_keyspace);
+	for(uint i = 0; i < graphs_in_keyspace_count; i ++) {
+		GraphContext_CreateMetaKeys(ctx, graphs_in_keyspace[i]);
+	}
+}
+
+// Module level function for RDB start end or load end event - Delete the meta keys for each graph.
+static void _ClearKeySpaceMetaKeys(RedisModuleCtx *ctx) {
+	uint graphs_in_keyspace_count = array_len(graphs_in_keyspace);
+	for(uint i = 0; i < graphs_in_keyspace_count; i ++) {
+		GraphContext_DeleteMetaKeys(ctx, graphs_in_keyspace[i]);
+	}
+}
+
+// Delete all meta keys before the actual flush, as the flush may be out of order
+static void _FlushDBHandler(RedisModuleCtx *ctx, RedisModuleEvent eid, uint64_t subevent,
+							void *data) {
+	if(eid.id == REDISMODULE_EVENT_FLUSHDB && subevent == REDISMODULE_SUBEVENT_FLUSHDB_START) {
+		_ClearKeySpaceMetaKeys(ctx);
+	}
+}
+
+// Checks if the event is persistence start event.
+static bool _IsEventPersistenceStart(RedisModuleEvent eid, uint64_t subevent) {
+	return eid.id == REDISMODULE_EVENT_PERSISTENCE  &&
+		   // Normal RDB.
+		   (subevent == REDISMODULE_SUBEVENT_PERSISTENCE_RDB_START ||
+			// preamble AOF.
+			subevent == REDISMODULE_SUBEVENT_PERSISTENCE_AOF_START ||
+			// SAVE and DEBUG RELOAD.
+			subevent == REDISMODULE_SUBEVENT_PERSISTENCE_SYNC_RDB_START);
+}
+
+// Checks if the event is persistence end event.
+static bool _IsEventPersistenceEnd(RedisModuleEvent eid, uint64_t subevent) {
+	return eid.id == REDISMODULE_EVENT_PERSISTENCE &&
+		   // Save ended.
+		   (subevent == REDISMODULE_SUBEVENT_PERSISTENCE_ENDED ||
+			// Save failed.
+			subevent == REDISMODULE_SUBEVENT_PERSISTENCE_FAILED);
+}
+
+// Checks if the event is loading end event.
+static bool _IsEventLoadingEnd(RedisModuleEvent eid, uint64_t subevent) {
+	return eid.id == REDISMODULE_EVENT_LOADING &&
+		   // Load ended.
+		   (subevent == REDISMODULE_SUBEVENT_LOADING_ENDED ||
+			// Load failed.
+			subevent == REDISMODULE_SUBEVENT_LOADING_FAILED);
+}
+
+// Server persistence event handler.
+static void _PersistenceEventHandler(RedisModuleCtx *ctx, RedisModuleEvent eid, uint64_t subevent,
+									 void *data) {
+	if(_IsEventPersistenceStart(eid, subevent)) _CreateKeySpaceMetaKeys(ctx);
+	else if(_IsEventPersistenceEnd(eid, subevent)) _ClearKeySpaceMetaKeys(ctx);
+}
+
+// Server loading event handler.
+static void _LoadingEventHandler(RedisModuleCtx *ctx, RedisModuleEvent eid, uint64_t subevent,
+								 void *data) {
+	if(_IsEventLoadingEnd(eid, subevent)) _ClearKeySpaceMetaKeys(ctx);
+}
+
+static void _RegisterServerEvents(RedisModuleCtx *ctx) {
```

Comment:
Let's move all the keyspace subscriptions to a dedicated file rather than overloading module.c with all of the functions invoked by this.